### PR TITLE
Redux

### PR DESCRIPTION
**Pull request by pannga**
New features like: 

Auto calculating
Auto fetching
Even better state managment.

I'm looking for bugs. 

I don't know if the localstorage is implemented correctly.

Check out demosite -> https://redux-hot-beans.herokuapp.com/

And thanks!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,11 +26,14 @@
...
     "react-copy-to-clipboard": "^5.0.2",
     "react-dom": "^16.13.1",
+    "react-redux": "^7.2.1",
     "react-router-dom": "^5.2.0",
     "react-scripts": "^3.4.1",
     "react-toastify": "^6.0.5",
     "redis": "^3.0.2",
-    "typescript": "^3.9.5",
+    "redux": "^4.0.5",
+    "redux-devtools-extension": "^2.13.8",
```
Comment (pawelrychlik) [inline package.json:38]: move it to devDependencies?

File: server.js [hunk 2]
```diff
@@ -148,6 +146,23 @@ app.get('/init', async (request, response) => {
   }
 });
 
+app.get('/customers/:sessionId', async (request, response) => {
+  const sessionId = request.params.sessionId;
+
+  try {
+    const customers = await Promise.all(
+      storeCustomers.map(async (customer) => {
```
Comment (pawelrychlik) [inline server.js:154]: no await, no need for async

File: src/components/Footer/index.js [hunk 3]
```diff
@@ -6,14 +6,13 @@ const Footer = () => {
       <div className="row">
         <div className="col-4">
           <p>
-            This demo store aims to showcase the functions of
-            Voucherify. This is not a real shop.
+            This demo store aims to showcase the functions of Voucherify. This
```
Comment (pawelrychlik) [inline src/components/Footer/index.js:9]: store (US) vs shop (UK), which one, then?

File: src/components/Sidebar/SidebarCampaigns.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    },
+  }))(MuiAccordionDetails);
+
+  return (
+    <>
+      <Accordion
+        square
+        key={campaign.name}
+        expanded={expanded === `${campaign.name}`}
+        onChange={handleChange(`${campaign.name}`)}
```
Comment (pawelrychlik) [inline src/components/Sidebar/SidebarCampaigns.js:63]: string interpolation not needed

File: src/redux/actions/cartActions.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    items,
+    paymentMethod
+  );
+  
+  // If voucher or promotion is not applied
+  if (_.isEmpty(discount)) {
+    checkoutPayload.source_id = getOrderId();
+    checkoutPayload.status = 'FULFILLED';
+
+    await sendPayload(checkoutPayload, 'order')
```
Comment (pawelrychlik) [inline src/redux/actions/cartActions.js:195]: either get rid of await and `.then()`s will handle the rest,
or get rid of `.then()`s - you're already awaiting `sendPayload()` result.

File: src/redux/actions/cartActions.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+      })
+      .catch((error) => {
+        toast.error('There was a problem with your purchase');
+        console.log('[checkoutCart]', error);
+      });
+    // If voucher is applied
+  } else if (discount.hasOwnProperty('code')) {
+    const discountCode = discount.code;
+    checkoutPayload.code = discountCode;
+    await sendPayload(checkoutPayload, 'redeem')
```
Comment (pawelrychlik) [inline src/redux/actions/cartActions.js:208]: await OR then()s

File: src/redux/actions/cartActions.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+      })
+      .catch((error) => {
+        toast.error('There was a problem with your purchase');
+        console.log('[checkoutCart][Discount]', error);
+      });
+    // If promotion is applied
+  } else if (discount.hasOwnProperty('banner')) {
+    const promotionId = discount.id;
+    checkoutPayload.promotionId = promotionId;
+    await sendPayload(checkoutPayload, 'redeem')
```
Comment (pawelrychlik) [inline src/redux/actions/cartActions.js:221]: await OR then()s

File: src/redux/actions/shopActions.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  return { type: GET_PRODUCTS_SUCCESS, payload: { products } };
+};
+
+export const getProductsError = () => {
+    return { type: GET_PRODUCTS_ERROR };
+  };
+
+export const getProducts = () => async (dispatch) => {
+  dispatch(getProductsRequest());
+  await fetch(`${process.env.REACT_APP_API_URL || ''}/products`, {
```
Comment (pawelrychlik) [inline src/redux/actions/shopActions.js:18]: await OR then()s

File: src/redux/actions/userActions.js [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+  return { type: GET_CUSTOMER_ERROR };
+};
+
+export const setPaymentMethod = (paymentMethod) => {
+  return { type: SET_PAYMENT_METHOD, payload: { paymentMethod } };
+};
+
+export const init = () => async (dispatch) => {
+  dispatch(initRequest());
+  await fetch(`${process.env.REACT_APP_API_URL || ''}/init`, {
```
Comment (pawelrychlik) [inline src/redux/actions/userActions.js:100]: await OR then()s

File: src/redux/actions/userActions.js [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    if (type === 'normal') {
+      dispatch(getCustomerSuccess(selectedCustomer));
+    } else {
+      const oldSelectedCustomer = getState().userReducer.selectedCustomer;
+      if (
+        oldSelectedCustomer.summary.orders.total_amount ===
+        selectedCustomer.summary.orders.total_amount
+      ) {
+        // If true -> wait
+        sleep(5000).then(() => dispatch(getCustomer(id, 'update')));
```
Comment (pawelrychlik) [inline src/redux/actions/userActions.js:237]: why not
```
await sleep(5000);
dispatch(...)
```

**Review by pawelrychlik — APPROVED** [review]
👏 😻

File: server.js (hunk 2 above)
Comment (pannga) [reply server.js:154]: I'm not sure why - but if I delete this, then everything stops (I'm waiting forever for customers). There is not any error, but I'm still waiting.

File: src/components/Footer/index.js (hunk 3 above)
Comment (pannga) [reply src/components/Footer/index.js:9]: Fixd

File: src/redux/actions/cartActions.js (hunk 5 above)
Comment (pannga) [reply src/redux/actions/cartActions.js:195]: Fixd. Thanks to this I found some other bug

File: src/redux/actions/userActions.js (hunk 9 above)
Comment (pannga) [reply src/redux/actions/userActions.js:100]: fixd

File: src/redux/actions/shopActions.js (hunk 8 above)
Comment (pannga) [reply src/redux/actions/shopActions.js:18]: fixd

File: src/redux/actions/cartActions.js (hunk 7 above)
Comment (pannga) [reply src/redux/actions/cartActions.js:221]: fixd

File: src/redux/actions/userActions.js (hunk 10 above)
Comment (pannga) [reply src/redux/actions/userActions.js:237]: Fixd

File: src/components/Sidebar/SidebarCampaigns.js (hunk 4 above)
Comment (pannga) [reply src/components/Sidebar/SidebarCampaigns.js:63]: fixd

File: package.json (hunk 1 above)
Comment (pannga) [reply package.json:38]: Moved

File: src/redux/actions/cartActions.js (hunk 6 above)
Comment (pannga) [reply src/redux/actions/cartActions.js]: fixd

**Pull request merged**